use vcr for tests

sckott opened the pull request:
[sc-9769]

- I had to change the tests that used `sample` to select a random set of things to test since you can't reuse a fixture/cassette to test against if you change the inputs that affect the http request matching. you could just add more tests to cover the cases you want to cover?
- api key is hidden via `key=<<votesmart_key>>` in test fixtures
- `tests/testthat/helper-vcr.R` is the vcr config file. note the `preserve_exact_body_bytes = TRUE` setting uses base64 response body encoding, and the `filter_sensitive_data` param is where we hide the key

thoughts? any questions?

sckott committed:
use vcr for tests, first draft [sc-9769]

aedobbyn reviewed (approved):
awesome, thanks for doing this!

aedobbyn commented:
i think sample works!

and wow so much faster!! 🚀

sckott commented:
> i think sample works!

Hmm, not sure we're on the same page here cause this PR has those lines commented out.  You were using `sample()` to get some set of inputs to pass to http requests - this doesn't jive with vcr because you have to make the same http requests as have been done in the fixtures. So I have commented out the `sample()` lines and just used a static set of inputs to http requests. see e.g., here https://github.com/decktools/votesmart/pull/26/files#diff-e5cc3a9ed2bcea267a6e9e28835871fede8e5b93eb0813cbbe5fa0f280d7b4f5

Thoughts?

aedobbyn commented:
ah ok is see. i think it's fine to not use a sample. it looks like there are still some cases where we are using one, though e.g. [here](https://github.com/decktools/votesmart/pull/26/files#diff-e5cc3a9ed2bcea267a6e9e28835871fede8e5b93eb0813cbbe5fa0f280d7b4f5R18). was that intentional?

sckott commented:
I think i just commented out the uses of `sample` that caused tests to fail on the 2nd or subsequent runs. The other uses of sample seem to be fine, though I just looked at `measure_get_measures` and it seems to be passing when it should not be b/c it's still using sample, so each time its run the number would be different, don't have time right now to look in to this.

aedobbyn commented:
ok no rush! i think as long as it's passing that's fine

sckott committed:
fix fixtures and remove sample() usage in tests

aedobbyn commented:
awesome!

yeah if possible i'd say we should add the api key and run the tests on CI. the key is in `secrets.get("VOTESMART_API_KEY")`

sckott commented:
Okay, will do, i'll report back after trying it

sckott committed:
populate vs api key in gh actions env, run tests on gh actions

sckott committed:
add --no-build-vignettes to args and add build_args with no build vign

sckott committed:
i think latest errors on CI are from a gestalt version thats too new

sckott committed:
rcmdcheck error on error only instead of warnings

sckott commented:
Okay, actions are succeeding now, okay with these changes?:

- added to rcmdcheck to error on an error only - instead of on any warnings
- installing gestalt v0.1.9 specifically on CI - nasty errors before doing that
- now that key is avail in CI, i added no-build-vignettes b/c it wanted to try to build and was throwing errors.

aedobbyn commented:
sweet!

> added to rcmdcheck to error on an error only - instead of on any warnings

do you know what the warnings we were getting were?

> installing gestalt v0.1.9 specifically on CI - nasty errors before doing that

do you think we should pin this version in the `DESCRIPTION`?

> now that key is avail in CI, i added no-build-vignettes b/c it wanted to try to build and was throwing errors.

ah the key being available there was causing the vignettes to not build?

aedobbyn commented:
so i think the `gestalt` issue should be taken care of with https://github.com/decktools/votesmart/pull/28

as far as the vignette, changing `knitr::rmarkdown` to `knitr::knitr` might fix the warning (according to [this](https://stackoverflow.com/questions/42670800/r-check-warning-files-in-the-vignettes-directory-but-no-files-in-inst-doc)) but if it doesn't i'm fine with having rcmdcheck error on an error only

i also got that [note](https://github.com/decktools/votesmart/actions/runs/4704025868/jobs/8343221892#step:8:149) when running `devtools::check` locally, i think because we don't define `election_state_id` anywhere in the global environment, it's a field returned from the api. since i think that would present an issue with getting a new version onto cran i'll add that to the `globalVariables` in `globals.R` in #28

sckott committed:
Merge branch 'main' into scott/use-vcr

sckott committed:
remove gestalt install on gh actions

sckott committed:
woops, also remove pak as it was only needed for gestalt install

sckott committed:
use knitr::knitr for vignette engine instead of knitr::rmarkdown

sckott committed:
try removing --no-manual

sckott committed:
removing no manual did not help

sckott committed:
revert vignette engine back to knitr::rmarkdown

sckott commented:
Okay, I: 
- removed the `gestalt` install from the gh action file
- tried getting rid of the warning about no inst/doc, but nothing I tried worked. that's not to say that something will solve it though, but I haven't found it yet. i tried: removing `--no-manual` and changing the vignette engine from `knitr::rmarkdown` to `knitr::knitr`. 

Okay with error on error only?

sckott committed:
Merge branch 'main' into scott/use-vcr

aedobbyn commented:
ok thanks for checking it out! i'm good with error on error only 👍
